File: src/FirstBadVersion_278.java
public class FirstBadVersion_278 {
    /* The isBadVersion API is defined in the parent class VersionControl.
      boolean isBadVersion(int version); */

    static class VersionControl {
        boolean isBadVersion(int version) {
            return version >= 4;
        }
    }
    private static class Solution extends VersionControl {
        public int firstBadVersion(int n) {
            int min = 1;
            int max = n;
            while (min < max) {
                int middle = min + (max - min) / 2;
                if (isBadVersion(middle)) {
                    max = middle;
                } else {
                    min = middle + 1;
                }
            }
            return min;
        }


        public int lastGoodVersion(int n) {
            int min = 0;
            int max = n;
            while (min < max) {
                int middle = min + (max - min) / 2;
                if (isGoodVersion(middle)) {
                    min = middle;
                } else {
                    max = middle - 1;
                }
            }
            return min;
        }

        private boolean isGoodVersion(int n) {
            return ! isBadVersion(n);
        }
    }

    public static void main(String[] args) {
        System.out.println(new Solution().firstBadVersion(12));
        System.out.println(new Solution().lastGoodVersion(12));
    }
}

File: src/FirstBadVersion_278.kt
/* The isBadVersion API is defined in the parent class VersionControl.
      fun isBadVersion(version: Int) : Boolean {} */
class FirstBadVersionSolutionKt {
    abstract class VersionControl(private val badVersion: Int) {
        fun isBadVersion(n: Int) = n >= badVersion
        abstract fun firstBadVersion(n: Int): Int
    }
    class Solution(badVersion: Int): VersionControl(badVersion) {
        override fun firstBadVersion(n: Int) : Int {
            var min = 1
            var max = n
            var middle: Int
            while (min < max) {
                middle = min + (max - min) / 2
                if (isBadVersion(middle)) {
                    max = middle
                } else {
                    min = middle + 1
                }
            }
            return min
        }
    }
}


fun main() {
    println(FirstBadVersionSolutionKt.Solution(4).firstBadVersion(12))
}